### ISSUE TITLE
Unified Memory Optimizations: Native Memory Reservation and GPU-side Augmentations

### ISSUE THREAD

**Issue by Vishwakarma-Atul** [ultralytics/ultralytics#23988]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Is your feature request related to a problem? Please describe.
Currently, users on high-performance Unified Memory systems (like Apple Silicon M-series) face two main bottlenecks when training large batches:

1. **OS Resource Contention**: There is no native way to "reserve" or "wire" memory to the Python process. This often leads to the OS compressing or swapping training memory to the SSD, significantly decreasing training and increasing latency.

2. **CPU Augmentation Bottleneck**: While cache='ram' exists, augmentations are still computed on the CPU by default. On Unified Memory systems, moving these computations to the GPU (MPS/CUDA) would be more efficient as it eliminates CPU-GPU transfer overhead and utilizes parallel GPU cores for math-heavy transforms.

### Describe the solution you'd like
I would like to see two new arguments added to the model.train() configuration:

1. **reserve_memory (bool/int):** * If True, the trainer should attempt to "wire" (lock) the process memory to physical RAM using mlockall (on Linux/macOS).

This prevents the OS from swapping training data to disk, ensuring peak performance.

2. **gpu_augment (bool):** * When enabled, the augmentation pipeline (flipping, rotation, color jitter, etc.) should be executed on the target device (MPS or CUDA) rather than the CPU.

This would allow the DataLoader to focus solely on disk I/O while the GPU handles both the math for augmentations and the training itself.

### Describe alternatives you've considered
**Manual Wrappers:** Using ctypes to call libc.mlockall manually in a script, but this is inaccessible for CLI users and requires root privileges that could be handled more gracefully by the framework.

**Custom Callbacks:** Implementing GPU-side augmentation via on_train_batch_start callbacks, but this is complex for most users and doesn't integrate natively with the YOLO hyperparameter YAMLs.

### Additional context
On an M5 Pro with Unified Memory, the physical RAM is shared. Currently, the CPU prepares a batch (taking 100% of several cores) and then "hands it" to the GPU. By moving this to the GPU, we can:

- Reduce total training time.
- Keep the CPU cooler/available for other tasks.
- Optimize for the "Zero-Copy" nature of modern silicon.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Vishwakarma-Atul, thank you for your interest in Ultralytics 🚀! This is an automated response to help get your feature request triaged quickly, and an Ultralytics engineer will also assist soon.
We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.
If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.
If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).
For this ✨ Feature Request, thanks for clearly describing the motivation, proposed API, and unified memory use case. To help the team evaluate it, any additional benchmark details, profiling results, or comparisons across devices and training settings would be helpful.
Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.
## Upgrade
Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:
```bash
pip install -U ultralytics
```
## Environments
YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>
## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
> If True, the trainer should attempt to "wire" (lock) the process memory to physical RAM using mlockall (on Linux/macOS).

`mlockall` [isn't supported on macOS](https://openbao.org/docs/rfcs/mlock-removal/) and locking memory in most cases requires eleveated permissions. So this is something we don't wish to implement.

> When enabled, the augmentation pipeline (flipping, rotation, color jitter, etc.) should be executed on the target device (MPS or CUDA) rather than the CPU.

The augmentation pipeline uses numpy arrays and OpenCV which can't use MPS, and it would require a very significant rewrite of whole pipeline to support that.

**Comment by Vishwakarma-Atul** [comment]
Has anyone audited yolo-mlx. from the documentation it look like similar interface as ultralytics but instead of pytorch-mps backend it uses mlx for training and inference.

https://github.com/thewebAI/yolo-mlx

please let me know if there are different mlx based librarie there which are as mature as ultralytics.

**Comment by glenn-jocher** [comment]
We haven't audited `yolo-mlx`, so I can't vouch for it directly. I’m also not aware of another MLX-native detection stack at Ultralytics maturity right now; the official [MLX Examples](https://github.com/ml-explore/mlx-examples) project is broad, but its published examples currently cover areas like classification, SAM, CLIP, and LLaVA rather than a full YOLO/object-detection training stack, and `thewebAI/yolo-mlx` itself is still very new. ([github.com](https://github.com/ml-explore/mlx-examples))
